src/topics/lda_model.cpp

```diff
@@ -16,30 +17,23 @@ lda_model::lda_model(const learn::dataset& docs, std::size_t num_topics)
     /* nothing */
 }
 
-void lda_model::save_doc_topic_distributions(const std::string& filename) const
+void lda_model::save_doc_topic_distributions(std::ostream& stream) const
 {
-    std::ofstream file{filename};
-    for (const auto& doc : docs_)
+    io::packed::write(stream, docs_.size());
+    io::packed::write(stream, num_topics_);
+
+    for (const auto& d : docs_)
     {
-        file << doc.id << "\t";
-        double sum = 0;
+        auto dist = topic_distrbution(doc_id{d.id});
         for (topic_id j{0}; j < num_topics_; ++j)
         {
-            double prob = compute_doc_topic_probability(doc.id, j);
-            if (prob > 0)
-                file << j << ":" << prob << "\t";
-            sum += prob;
+            io::packed::write(stream, dist.counts(j));
         }
-        if (std::abs(sum - 1) > 1e-6)
-            throw std::runtime_error{"invalid probability distribution"};
-        file << "\n";
     }
 }
 
-void lda_model::save_topic_term_distributions(const std::string& filename) const
+void lda_model::save_topic_term_distributions(std::ostream& stream) const
```

Comment:
Where you have

```cpp
sum *= (1 / model.num_topics());
```

you probably want

```cpp
sum *= (1.0 / model.num_topics());
```

as I think what's happening may be `sum *= 0;` because of integer math. Hopefully that fixes it?

(I also like the `bl_term_scorer` name.)